oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderService.java

```diff
@@ -209,45 +208,51 @@ private void initializeExtractedTextCache(Map<String, ?> config, StatisticsProvi
-                "directory path [%s] nor repository home [%s] defined", PROP_LOCAL_TEXT_EXTRACTION_DIR, REPOSITORY_HOME);
+        if (textExtractionDir == null) {
+            throw new IllegalStateException(String.format("Text extraction directory cannot be determined as neither " +
+                    "directory path [%s] nor repository home [%s] defined", PROP_LOCAL_TEXT_EXTRACTION_DIR, REPOSITORY_HOME));
+        }
 
         this.textExtractionDir = new File(textExtractionDir);
     }
 
-    private void registerExtractedTextProvider(PreExtractedTextProvider provider){
-        if (extractedTextCache != null){
-            if (provider != null){
+    private void registerExtractedTextProvider(PreExtractedTextProvider provider) {
+        if (extractedTextCache != null) {
+            if (provider != null) {
                 String usage = extractedTextCache.isAlwaysUsePreExtractedCache() ?
                         "always" : "only during reindexing phase";
                 LOG.info("Registering PreExtractedTextProvider {} with extracted text cache. " +
-                        "It would be used {}",  provider, usage);
+                        "It would be used {}", provider, usage);
             } else {
                 LOG.info("Unregistering PreExtractedTextProvider with extracted text cache");
             }
             extractedTextCache.setExtractedTextProvider(provider);
         }
     }
 
-    private ElasticsearchIndexCoordinateFactory getElasticsearchIndexCoordinateFactory(Map<String, ?> config) {
-        ElasticsearchIndexCoordinateFactory esIndexCoordFactory;
-        Map<String, String> esCfg = Maps.newHashMap();
-        esCfg.put(ElasticsearchCoordinate.SCHEME_PROP,
-                PropertiesUtil.toString(config.get(PROP_ELASTICSEARCH_SCHEME), PROP_ELASTICSEARCH_SCHEME_DEFAULT));
-        esCfg.put(ElasticsearchCoordinate.HOST_PROP,
-                PropertiesUtil.toString(config.get(PROP_ELASTICSEARCH_HOST), PROP_ELASTICSEARCH_HOST_DEFAULT));
-        esCfg.put(ElasticsearchCoordinate.PORT_PROP, String.valueOf(
-                PropertiesUtil.toInteger(config.get(PROP_ELASTICSEARCH_PORT), PROP_ELASTICSEARCH_PORT_DEFAULT)));
-        esIndexCoordFactory = new DefaultElasticsearchIndexCoordinateFactory(connectionFactory, esCfg);
-        return esIndexCoordFactory;
+    private ElasticsearchCoordinate getElasticsearchCoordinate(Map<String, ?> contextConfig) {
+        // system properties have priority
+        ElasticsearchCoordinate coordinate = ElasticsearchCoordinate.build(System.getProperties().entrySet()
+                .stream()
+                .collect(Collectors.toMap(
+                        e -> String.valueOf(e.getKey()),
+                        e -> String.valueOf(e.getValue()))
+                )
+        );
+
+        if (coordinate == null) {
+            coordinate = ElasticsearchCoordinate.build(contextConfig);
```

Comment:
I see now why you return null in build... but I think it's better to have the logic in the caller than in build.